packages/dds/tree/docs/wip/Schema.ts

```diff
@@ -0,0 +1,74 @@
+/**
+ * Example internal schema representation types.
+ */
+
+/**
+ * Stable identifier, used when persisting data.
+ *
+ * This intentionally does not support any concept of versioning or name-spacing:
+ * users of it can include versions and namespaces in if they want, or features for them could be added later.
+ *
+ * Can be used either:
+ * 1. Record just this in the persisted data.
+ * When loading the data, see if the loader has a schema with a matching name, and if so, use that.
+ * Optionally validate loaded data against schema.
+ * 2. Persist the whole schema. Use the identifier to associate it with schema when loading to check that the schema match.
+ */
+export type SchemaIdentifier = string;
+
+export enum Multiplicity {
+    /**
+     * Exactly one item.
+     */
+    Value,
+    /**
+     * 0 or 1 items.
+     */
+    Optional,
+    /**
+     * 0 or more items.
+     */
+    Sequence,
+    /**
+     * Exactly 0 items.
+     */
+    Forbidden,
+}
+
+export enum Value {
+    Nothing,
+    Number,
+    String,
+    Boolean,
+    /**
+     * Any Fluid serializable data.
+     */
+    Serializable,
+}
+
+export interface FieldContent {
+    multiplicity: Multiplicity;
+    /**
+     * If not specified, child types are unconstrained.
+     */
+    types?: Type[];
+}
+
+export interface Field {
+    content: FieldContent;
+    /**
+     * Stable identifier, used when persisting data.
+     *
+     * Can be used either:
+     * 1. Record just this in the persisted data. When loading the data, see if the loader has a schema with a matching name, and if so, use that.
+     * 2. Persist the whole schema. Use the identifier to associate it with schema when loading to check that the schema match.
+     */
+    name: string;
+}
+
+export interface Type {
+    name: string;
+    fields: Field[];
+    extraFields?: FieldContent;
```

Comment:
To confirm my understanding of the semantics:
- when `extraFields` is missing then the type is a closed record. It has exactly the members specified by fields.
- when `extraFields` is present, then any field not in `fields` has the type specified by `extraFields`.

Are fields scoped? That is, do we assume that fields can have duplicate names but the type assigned to the field in the tree is the first? For example, is the following legal?
```
export const string: Type = {
    name: "foo",
    fields: [
        {
            name: "x",
            content: {
                multiplicity: Multiplicity.Value,
                types: [number],
            },
        },
        {
            name: "x",
            content: {
                multiplicity: Multiplicity.Value,
                types: [boolean],
            },
        },
    ],
    value: Value.Nothing,
};
```

Reply:
Yes, that is the semantics.

No, duplicate field names are not permitted. Fields are identified only by their node+key (name got renamed to key), and there is at most one field with a given key (well, there is a caveat that global and local keys are considered in separate namespaces and thus always distinct): this is true both of the types/schema and the nodes themselves. The schema now use sets which should reduce confusion over this.

The example code has been updated to explicitly address scoping, and majorly rework extraFields.